File: src-tauri/src/services/resource_indexer.rs
use regex::Regex;
use serde::{Deserialize, Serialize};
use std::collections::{HashMap, HashSet};
use std::fs;
use std::io::{BufRead, BufReader};
use std::path::{Path, PathBuf};

const CACHE_FILE: &str = "session-resources.json";

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Resource {
    #[serde(rename = "type")]
    pub resource_type: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub url: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub repo: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub state: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct CacheEntry {
    resources: Vec<Resource>,
    #[serde(default)]
    manual_resources: Option<Vec<Resource>>,
    #[serde(default)]
    removed_keys: Option<Vec<String>>,
    #[serde(rename = "indexedAt")]
    indexed_at: u128,
}

pub struct ResourceIndexer {
    session_state_dir: PathBuf,
    cache_path: PathBuf,
    cache: HashMap<String, CacheEntry>,
}

fn resource_key(r: &Resource) -> String {
    if let Some(ref id) = r.id {
        return format!("{}:{}", r.resource_type, id);
    }
    let url = r.url.as_deref().unwrap_or("");
    format!("{}:{}", r.resource_type, url)
}

impl ResourceIndexer {
    pub fn new(session_state_dir: &Path) -> Self {
        let cache_path = session_state_dir.join(CACHE_FILE);
        let mut indexer = Self {
            session_state_dir: session_state_dir.to_path_buf(),
            cache_path,
            cache: HashMap::new(),
        };
        indexer.load_cache();
        indexer.rebuild_if_stale();
        indexer
    }

    fn load_cache(&mut self) {
        if let Ok(data) = fs::read_to_string(&self.cache_path) {
            if let Ok(cache) = serde_json::from_str(&data) {
                self.cache = cache;
            }
        }
    }

    fn save_cache(&self) {
        if let Ok(json) = serde_json::to_string_pretty(&self.cache) {
            let _ = fs::write(&self.cache_path, json);
        }
    }

    fn rebuild_if_stale(&mut self) {
        let entries = match fs::read_dir(&self.session_state_dir) {
            Ok(e) => e,
            Err(_) => return,
        };

        let current_ids: HashSet<String> = entries
            .flatten()
            .filter(|e| e.file_type().map(|ft| ft.is_dir()).unwrap_or(false))
            .map(|e| e.file_name().to_string_lossy().to_string())
            .collect();

        let mut updated = false;

        self.cache.retain(|id, _| {
            let keep = current_ids.contains(id);
            if !keep { updated = true; }
            keep
        });

        for session_id in &current_ids {
            let session_dir = self.session_state_dir.join(session_id);

            if let Some(cached) = self.cache.get(session_id) {
                if let Ok(stat) = fs::metadata(&session_dir) {
                    let mtime = stat.modified().ok()
                        .and_then(|t| t.duration_since(std::time::SystemTime::UNIX_EPOCH).ok())
                        .map(|d| d.as_millis())
                        .unwrap_or(0);
                    if mtime <= cached.indexed_at {
                        continue;
                    }
                }
            }

            let resources = self.extract_resources(&session_dir);
            let existing = self.cache.remove(session_id);
            self.cache.insert(session_id.clone(), CacheEntry {
                resources,
                manual_resources: existing.as_ref().and_then(|e| e.manual_resources.clone()),
                removed_keys: existing.as_ref().and_then(|e| e.removed_keys.clone()),
                indexed_at: std::time::SystemTime::now()
                    .duration_since(std::time::SystemTime::UNIX_EPOCH)
                    .map(|d| d.as_millis())
                    .unwrap_or(0),
            });
            updated = true;
        }

        if updated {
            self.save_cache();
        }
    }

    fn extract_resources(&self, session_dir: &Path) -> Vec<Resource> {
        let events_path = session_dir.join("events.jsonl");
        let file = match fs::File::open(&events_path) {
            Ok(f) => f,
            Err(_) => return Vec::new(),
        };

        let reader = BufReader::new(file);
        let mut prs: HashMap<String, Resource> = HashMap::new();
        let mut work_items: HashMap<String, Resource> = HashMap::new();
        let mut repos: HashSet<String> = HashSet::new();
        let mut wiki_urls: HashSet<String> = HashSet::new();
        let mut pipelines: HashMap<String, Resource> = HashMap::new();

        let pr_url_re = Regex::new(r#"https?://[^\s"\\)]*?/pullrequest/(\d+)"#).unwrap();
        let wi_url_re = Regex::new(r#"https?://[^\s"\\)]*?/_workitems/edit/(\d+)"#).unwrap();
        let git_url_re = Regex::new(r#"https?://(?:microsoft\.visualstudio\.com|dev\.azure\.com/microsoft)/(?:DefaultCollection/)?(\w+)/_git/([^\s"\\);,]+)"#).unwrap();
        let wiki_re = Regex::new(r#"https?://[^\s"\\)]*?/_wiki/[^\s"\\)]+"#).unwrap();
        let pipeline_re = Regex::new(r#"https?://[^\s"\\)]*?/_build/results\?buildId=(\d+)[^\s"\\)]*"#).unwrap();
        let pr_id_re = Regex::new(r#""pullRequestId"\s*:\s*(\d+)"#).unwrap();
        let wi_id_re = Regex::new(r#""workItemId"\s*:\s*(\d+)"#).unwrap();
        let pr_repo_re = Regex::new(r"_git/([^/]+)/pullrequest").unwrap();

        for line in reader.lines().flatten() {
            // PR URLs
            for cap in pr_url_re.captures_iter(&line) {
                let pr_id = cap[1].to_string();
                let url = cap[0].trim_end_matches(&['"', ')', '}', ']', '\\'][..]).to_string();
                let repo_match = pr_repo_re.captures(&url)
                    .map(|c| c[1].to_string());
                prs.entry(pr_id.clone()).or_insert(Resource {
                    resource_type: "pr".to_string(),
                    id: Some(pr_id),
                    url: Some(url),
                    name: None,
                    repo: repo_match,
                    state: None,
                });
            }

            // PR IDs from tool calls
            for cap in pr_id_re.captures_iter(&line) {
                let pr_id = cap[1].to_string();
                prs.entry(pr_id.clone()).or_insert(Resource {
                    resource_type: "pr".to_string(),
                    id: Some(pr_id),
                    url: None, name: None, repo: None, state: None,
                });
            }

            // Work item URLs
            for cap in wi_url_re.captures_iter(&line) {
                let wi_id = cap[1].to_string();
                let url = cap[0].trim_end_matches(&['"', ')', '}', ']', '\\'][..]).to_string();
                work_items.entry(wi_id.clone()).or_insert(Resource {
                    resource_type: "workitem".to_string(),
                    id: Some(wi_id), url: Some(url),
                    name: None, repo: None, state: None,
                });
            }

            // Work item IDs from tool calls
            for cap in wi_id_re.captures_iter(&line) {
                let wi_id = cap[1].to_string();
                work_items.entry(wi_id.clone()).or_insert(Resource {
                    resource_type: "workitem".to_string(),
                    id: Some(wi_id), url: None,
                    name: None, repo: None, state: None,
                });
            }

            // Git repo URLs (exclude PR URLs)
            for cap in git_url_re.captures_iter(&line) {
                let mut url = cap[0].trim_end_matches(&['"', ')', '}', ']', '\\', ';', ','][..]).to_string();
                if !url.contains("/pullrequest/") {
                    if let Some(git_idx) = url.find("/_git/") {
                        let after = &url[git_idx + 6..];
                        let repo_name = after.split('/').next().unwrap_or(after);
                        url = format!("{}{}", &url[..git_idx + 6], repo_name);
                    }
                    repos.insert(url);
                }
            }

            // Wiki URLs
            for cap in wiki_re.captures_iter(&line) {
                wiki_urls.insert(cap[0].trim_end_matches(&['"', ')', '}', ']', '\\'][..]).to_string());
            }

            // Pipeline URLs
            for cap in pipeline_re.captures_iter(&line) {
                let build_id = cap[1].to_string();
                let url = cap[0].trim_end_matches(&['"', ')', '}', ']', '\\'][..]).to_string();
                pipelines.entry(build_id.clone()).or_insert(Resource {
                    resource_type: "pipeline".to_string(),
                    id: Some(build_id), url: Some(url),
                    name: None, repo: None, state: None,
                });
            }
        }

        let mut resources: Vec<Resource> = Vec::new();
        resources.extend(prs.into_values());
        resources.extend(work_items.into_values());
        let repo_name_re = Regex::new(r"_git/(.+)").unwrap();
        for url in repos {
            let name = repo_name_re.captures(&url)
                .map(|c| c[1].to_string())
                .unwrap_or_else(|| url.clone());
            resources.push(Resource {
                resource_type: "repo".to_string(),
                id: None, url: Some(url), name: Some(name),
                repo: None, state: None,
            });
        }
        for url in wiki_urls {
            resources.push(Resource {
                resource_type: "wiki".to_string(),
                id: None, url: Some(url), name: None,
                repo: None, state: None,
            });
        }
        resources.extend(pipelines.into_values());

        resources
    }

    pub fn get_resources_for_session(&self, session_id: &str) -> Vec<serde_json::Value> {
        let entry = match self.cache.get(session_id) {
            Some(e) => e,
            None => return Vec::new(),
        };

        let auto = &entry.resources;
        let manual = entry.manual_resources.as_deref().unwrap_or(&[]);
        let removed: HashSet<String> = entry.removed_keys.as_ref()
            .map(|v| v.iter().cloned().collect())
            .unwrap_or_default();

        let mut seen = HashSet::new();
        let mut merged = Vec::new();

        for r in auto.iter().chain(manual.iter()) {
            let key = resource_key(r);
            if removed.contains(&key) || seen.contains(&key) {
                continue;
            }
            seen.insert(key);
            if let Ok(val) = serde_json::to_value(r) {
                merged.push(val);
            }
        }

        merged
    }

    /// Parse a URL string into a typed Resource.
    fn parse_url_to_resource(url: &str) -> Option<Resource> {
        let url = url.trim();
        if url.is_empty() { return None; }

        if url.contains("/pullrequest/") {
            let pr_re = Regex::new(r"/pullrequest/(\d+)").ok()?;
            let id = pr_re.captures(url)?.get(1)?.as_str().to_string();
            let repo = Regex::new(r"_git/([^/]+)/pullrequest").ok()
                .and_then(|re| re.captures(url))
                .map(|c| c[1].to_string());
            return Some(Resource {
                resource_type: "pr".to_string(),
                id: Some(id), url: Some(url.to_string()),
                repo, name: None, state: None,
            });
        }
        if url.contains("/_workitems/edit/") {
            let re = Regex::new(r"/_workitems/edit/(\d+)").ok()?;
            let id = re.captures(url)?.get(1)?.as_str().to_string();
            return Some(Resource {
                resource_type: "workitem".to_string(),
                id: Some(id), url: Some(url.to_string()),
                name: None, repo: None, state: None,
            });
        }
        if url.contains("/_build/results?buildId=") {
            let re = Regex::new(r"buildId=(\d+)").ok()?;
            let id = re.captures(url)?.get(1)?.as_str().to_string();
            return Some(Resource {
                resource_type: "pipeline".to_string(),
                id: Some(id), url: Some(url.to_string()),
                name: None, repo: None, state: None,
            });
        }
        if url.contains("/_build?definitionId=") {
            let re = Regex::new(r"definitionId=(\d+)").ok()?;
            let def_id = format!("def-{}", re.captures(url)?.get(1)?.as_str());
            return Some(Resource {
                resource_type: "pipeline".to_string(),
                id: Some(def_id), url: Some(url.to_string()),
                name: None, repo: None, state: None,
            });
        }
        if url.contains("releaseId=") {
            let re = Regex::new(r"releaseId=(\d+)").ok()?;
            let id = re.captures(url)?.get(1)?.as_str().to_string();
            return Some(Resource {
                resource_type: "release".to_string(),
                id: Some(id), url: Some(url.to_string()),
                name: None, repo: None, state: None,
            });
        }
        if url.contains("/_wiki/") {
            return Some(Resource {
                resource_type: "wiki".to_string(),
                id: None, url: Some(url.to_string()),
                name: None, repo: None, state: None,
            });
        }
        if url.contains("/_git/") {
            // Normalize to /_git/RepoName only
            let mut normalized = url.to_string();
            if let Some(git_idx) = normalized.find("/_git/") {
                let after = &normalized[git_idx + 6..];
                let repo_name = after.split('/').next().unwrap_or(after).to_string();
                normalized = format!("{}/_git/{}", &normalized[..git_idx], repo_name);
            }
            let name = Regex::new(r"_git/(.+)").ok()
                .and_then(|re| re.captures(&normalized))
                .map(|c| c[1].to_string())
                .unwrap_or_else(|| normalized.clone());
            return Some(Resource {
                resource_type: "repo".to_string(),
                id: None, url: Some(normalized),
                name: Some(name), repo: None, state: None,
            });
        }
        // Generic link
        let display_name = url.trim_start_matches("https://").trim_start_matches("http://")
            .split('/').take(3).collect::<Vec<_>>().join("/");
        Some(Resource {
            resource_type: "link".to_string(),
            id: None, url: Some(url.to_string()),
            name: Some(display_name), repo: None, state: None,
        })
    }

    /// Add a manually-specified resource (by URL) to a session.
    /// Returns `true` if added, `false` if it was a duplicate.
    pub fn add_manual_resource(&mut self, session_id: &str, url: &str) -> bool {
        let resource = match Self::parse_url_to_resource(url) {
            Some(r) => r,
            None => return false,
        };
        let entry = self.cache.entry(session_id.to_string()).or_insert_with(|| CacheEntry {
            resources: Vec::new(),
            manual_resources: None,
            removed_keys: None,
            indexed_at: 0,
        });
        let manual = entry.manual_resources.get_or_insert_with(Vec::new);
        let removed = entry.removed_keys.get_or_insert_with(Vec::new);
        let key = resource_key(&resource);

        // Check for duplicate across auto + manual
        let existing_keys: HashSet<String> = entry.resources.iter()
            .chain(manual.iter())
            .map(resource_key)
            .collect();
        if existing_keys.contains(&key) {
            return false;
        }

        // Un-remove if previously removed
        removed.retain(|k| k != &key);
        manual.push(resource);
        self.save_cache();
        true
    }

    /// Mark a resource as removed for a session (hides it from results).
    pub fn remove_resource(&mut self, session_id: &str, key: &str) {
        if let Some(entry) = self.cache.get_mut(session_id) {
            let removed = entry.removed_keys.get_or_insert_with(Vec::new);
            if !removed.contains(&key.to_string()) {
                removed.push(key.to_string());
            }
            // Also remove from manual list
            if let Some(manual) = entry.manual_resources.as_mut() {
                manual.retain(|r| resource_key(r) != key);
            }
            self.save_cache();
        }
    }
}

File: src/resource-indexer.js
const fs = require('fs');
const path = require('path');
const readline = require('readline');

const CACHE_FILE = 'session-resources.json';
const REBUILD_INTERVAL_MS = 10 * 60 * 1000;

// Patterns for extracting resources
const PR_URL_RE = /https?:\/\/[^\s"\\)]*\/pullrequest\/(\d+)/g;
const WI_URL_RE = /https?:\/\/[^\s"\\)]*\/_workitems\/edit\/(\d+)/g;
const GIT_URL_RE = /https?:\/\/(?:microsoft\.visualstudio\.com|dev\.azure\.com\/microsoft)\/(?:DefaultCollection\/)?(\w+)\/_git\/([^\s"\\);,]+)/g;
const WIKI_URL_RE = /https?:\/\/[^\s"\\)]*\/_wiki\/[^\s"\\)]+/g;
const PIPELINE_URL_RE = /https?:\/\/[^\s"\\)]*\/_build\/results\?buildId=(\d+)[^\s"\\)]*/g;
const PIPELINE_DEF_URL_RE = /https?:\/\/[^\s"\\)]*\/_build\?definitionId=(\d+)[^\s"\\)]*/g;
const RELEASE_URL_RE = /https?:\/\/[^\s"\\)]*\/_releaseProgress\?[^\s"\\)]*releaseId=(\d+)[^\s"\\)]*/g;
const PR_ID_TOOL_RE = /"pullRequestId"\s*:\s*(\d+)/g;
const PR_STATUS_RE = /"pullRequestId"\s*:\s*(\d+)[^}]*?"status"\s*:\s*"(active|completed|abandoned)"/gi;
const PR_STATUS_ESCAPED_RE = /\\?"pullRequestId\\?"\s*:\s*(\d+)[\s\S]{0,200}?\\?"status\\?"\s*:\s*(\d+|\\?"(?:active|completed|abandoned)\\?")/gi;
const WI_ID_TOOL_RE = /"workItemId"\s*:\s*(\d+)/g;

function normalizeRepoUrl(url) {
  url = (url || '').replace(/[?#].*$/, '').replace(/\/+$/, '');
  // Canonicalize host variants to a consistent form
  url = url.replace(/microsoft\.visualstudio\.com/, 'dev.azure.com/microsoft');
  url = url.replace(/\/DefaultCollection\//, '/');
  // Strip path segments after /_git/RepoName
  const gitIdx = url.indexOf('/_git/');
  if (gitIdx !== -1) {
    const afterGit = url.substring(gitIdx + 6);
    const repoName = afterGit.split('/')[0];
    url = url.substring(0, gitIdx + 6) + repoName;
  }
  return url;
}

function resourceKey(r) {
  if (r.id) return `${r.type}:${r.id}`;
  const url = r.type === 'repo' ? normalizeRepoUrl(r.url) : (r.url || '').replace(/[?#].*$/, '').replace(/\/+$/, '');
  return `${r.type}:${url}`;
}

class ResourceIndexer {
  constructor(sessionStateDir) {
    this.sessionStateDir = sessionStateDir;
    this.cachePath = path.join(sessionStateDir, CACHE_FILE);
    this.cache = {};
    this._rebuildTimer = null;
  }

  async init() {
    await this._loadCache();
    await this.rebuildIfStale();
    this._startPeriodicRebuild();
  }

  async _loadCache() {
    try {
      const data = await fs.promises.readFile(this.cachePath, 'utf8');
      this.cache = JSON.parse(data);
    } catch {
      this.cache = {};
    }
  }

  async _saveCache() {
    try {
      await fs.promises.writeFile(this.cachePath, JSON.stringify(this.cache, null, 2), 'utf8');
    } catch {}
  }

  _startPeriodicRebuild() {
    this._rebuildTimer = setInterval(() => this.rebuildIfStale(), REBUILD_INTERVAL_MS);
  }

  stop() {
    if (this._rebuildTimer) clearInterval(this._rebuildTimer);
  }

  async rebuildIfStale() {
    const entries = await fs.promises.readdir(this.sessionStateDir, { withFileTypes: true });
    const currentIds = new Set(entries.filter(e => e.isDirectory()).map(e => e.name));
    let updated = false;

    // Prune orphaned cache entries
    for (const id of Object.keys(this.cache)) {
      if (!currentIds.has(id)) { delete this.cache[id]; updated = true; }
    }

    for (const entry of entries) {
      if (!entry.isDirectory()) continue;
      const sessionId = entry.name;
      const sessionDir = path.join(this.sessionStateDir, sessionId);

      const cached = this.cache[sessionId];
      if (cached) {
        try {
          const stat = await fs.promises.stat(sessionDir);
          if (stat.mtime.getTime() <= cached.indexedAt) continue;
        } catch { continue; }
      }

      const resources = await this._extractResources(sessionDir);
      this.cache[sessionId] = { resources, indexedAt: Date.now() };
      updated = true;
    }

    if (updated) await this._saveCache();
  }

  async _extractResources(sessionDir) {
    const eventsPath = path.join(sessionDir, 'events.jsonl');
    try { await fs.promises.access(eventsPath); } catch { return []; }

    const prs = new Map();       // prId -> { id, url?, repo? }
    const workItems = new Map(); // wiId -> { id, url? }
    const repos = new Set();     // repo URLs
    const wikiUrls = new Set();
    const pipelines = new Map(); // buildId -> { id, url }
    const releases = new Map();  // releaseId -> { id, url }

    return new Promise((resolve) => {
      const stream = fs.createReadStream(eventsPath, { encoding: 'utf8' });
      const rl = readline.createInterface({ input: stream, crlfDelay: Infinity });

      rl.on('line', (line) => {
        // Search the raw line for efficiency (most patterns work across any field)
        let m;

        // PR URLs (includes PR ID + repo context)
        const prUrlRe = new RegExp(PR_URL_RE.source, 'g');
        while ((m = prUrlRe.exec(line)) !== null) {
          const prId = m[1];
          const url = m[0].replace(/[)}\]"\\]+$/, ''); // strip trailing junk
          const repoMatch = url.match(/_git\/([^/]+)\/pullrequest/);
          const existing = prs.get(prId);
          if (existing) {
            if (!existing.url) existing.url = url;
            if (!existing.repo && repoMatch) existing.repo = repoMatch[1];
          } else {
            prs.set(prId, {
              id: prId,
              url,
              repo: repoMatch ? repoMatch[1] : null,
              type: 'pr',
              state: null
            });
          }
        }

        // PR IDs from tool calls
        const prIdRe = new RegExp(PR_ID_TOOL_RE.source, 'g');
        while ((m = prIdRe.exec(line)) !== null) {
          const prId = m[1];
          if (!prs.has(prId)) {
            prs.set(prId, { id: prId, url: null, repo: null, type: 'pr', state: null });
          }
        }

        // PR status from tool call arguments (e.g. "pullRequestId": 123, ..., "status": "Active")
        const prStatusRe = new RegExp(PR_STATUS_RE.source, 'gi');
        while ((m = prStatusRe.exec(line)) !== null) {
          const prId = m[1];
          const state = m[2].toLowerCase();
          const existing = prs.get(prId);
          if (existing) {
            existing.state = state;
          }
        }

        // PR status from escaped JSON in tool responses (e.g. \"status\":1 or \"status\":\"active\")
        const prStatusEscRe = new RegExp(PR_STATUS_ESCAPED_RE.source, 'gi');
        while ((m = prStatusEscRe.exec(line)) !== null) {
          const prId = m[1];
          const rawState = m[2].replace(/\\?"/g, '');
          const STATUS_MAP = { '1': 'active', '2': 'abandoned', '3': 'completed', 'active': 'active', 'abandoned': 'abandoned', 'completed': 'completed' };
          const state = STATUS_MAP[rawState.toLowerCase()];
          if (state) {
            const existing = prs.get(prId);
            if (existing) existing.state = state;
          }
        }

        // Work item URLs
        const wiUrlRe = new RegExp(WI_URL_RE.source, 'g');
        while ((m = wiUrlRe.exec(line)) !== null) {
          const wiId = m[1];
          const url = m[0].replace(/[)}\]"\\]+$/, '');
          workItems.set(wiId, { id: wiId, url, type: 'workitem' });
        }

        // Work item IDs from tool calls
        const wiIdRe = new RegExp(WI_ID_TOOL_RE.source, 'g');
        while ((m = wiIdRe.exec(line)) !== null) {
          const wiId = m[1];
          if (!workItems.has(wiId)) {
            workItems.set(wiId, { id: wiId, url: null, type: 'workitem' });
          }
        }

        // Git repo URLs (exclude PR URLs)
        const gitUrlRe = new RegExp(GIT_URL_RE.source, 'g');
        while ((m = gitUrlRe.exec(line)) !== null) {
          let repoUrl = m[0].replace(/[)}\]"\\;,]+$/, '');
          if (!repoUrl.includes('/pullrequest/')) {
            // Normalize: keep only up to /_git/RepoName
            const gitIdx = repoUrl.indexOf('/_git/');
            if (gitIdx !== -1) {
              const afterGit = repoUrl.substring(gitIdx + 6);
              const repoName = afterGit.split('/')[0];
              repoUrl = repoUrl.substring(0, gitIdx + 6) + repoName;
            }
            repos.add(repoUrl);
          }
        }

        // Wiki URLs
        const wikiRe = new RegExp(WIKI_URL_RE.source, 'g');
        while ((m = wikiRe.exec(line)) !== null) {
          wikiUrls.add(m[0].replace(/[)}\]"\\]+$/, ''));
        }

        // Pipeline URLs (build results)
        const pipelineRe = new RegExp(PIPELINE_URL_RE.source, 'g');
        while ((m = pipelineRe.exec(line)) !== null) {
          const buildId = m[1];
          if (!pipelines.has(buildId)) {
            pipelines.set(buildId, { id: buildId, url: m[0].replace(/[)}\]"\\]+$/, ''), type: 'pipeline' });
          }
        }
        const pipelineDefRe = new RegExp(PIPELINE_DEF_URL_RE.source, 'g');
        while ((m = pipelineDefRe.exec(line)) !== null) {
          const defId = `def-${m[1]}`;
          if (!pipelines.has(defId)) {
            pipelines.set(defId, { id: defId, url: m[0].replace(/[)}\]"\\]+$/, ''), type: 'pipeline' });
          }
        }

        // Release URLs
        const releaseRe = new RegExp(RELEASE_URL_RE.source, 'g');
        while ((m = releaseRe.exec(line)) !== null) {
          const releaseId = m[1];
          if (!releases.has(releaseId)) {
            releases.set(releaseId, { id: releaseId, url: m[0].replace(/[)}\]"\\]+$/, ''), type: 'release' });
          }
        }
      });

      rl.on('close', () => {
        const resources = [
          ...[...prs.values()],
          ...[...workItems.values()],
          ...[...repos].map(url => {
            const repoName = url.match(/_git\/(.+)/)?.[1] || url;
            return { type: 'repo', name: repoName, url };
          }),
          ...[...wikiUrls].map(url => ({ type: 'wiki', url })),
          ...[...pipelines.values()],
          ...[...releases.values()]
        ];
        resolve(resources);
      });

      rl.on('error', () => resolve([]));
    });
  }

  getResourcesForSession(sessionId) {
    const entry = this.cache[sessionId];
    if (!entry) return [];
    const auto = entry.resources || [];
    const manual = entry.manualResources || [];
    const removed = new Set(entry.removedKeys || []);
    const seen = new Set();
    const merged = [];
    for (const r of [...auto, ...manual]) {
      const key = resourceKey(r);
      if (removed.has(key) || seen.has(key)) continue;
      seen.add(key);
      merged.push(r);
    }
    return merged;
  }

  async addManualResource(sessionId, resource) {
    if (!this.cache[sessionId]) {
      this.cache[sessionId] = { resources: [], indexedAt: 0 };
    }
    const entry = this.cache[sessionId];
    if (!entry.manualResources) entry.manualResources = [];
    if (!entry.removedKeys) entry.removedKeys = [];

    const key = resourceKey(resource);
    // Check for duplicates across auto + manual
    const allKeys = new Set([
      ...(entry.resources || []).map(resourceKey),
      ...entry.manualResources.map(resourceKey)
    ]);
    if (allKeys.has(key)) return { added: false, reason: 'duplicate' };

    // Un-remove if it was previously removed
    entry.removedKeys = entry.removedKeys.filter(k => k !== key);

    entry.manualResources.push(resource);
    await this._saveCache();
    return { added: true };
  }

  async removeResource(sessionId, key) {
    const entry = this.cache[sessionId];
    if (!entry) return;
    if (!entry.removedKeys) entry.removedKeys = [];
    if (!entry.removedKeys.includes(key)) entry.removedKeys.push(key);
    // Also remove from manual if it was manually added
    if (entry.manualResources) {
      entry.manualResources = entry.manualResources.filter(r => resourceKey(r) !== key);
    }
    await this._saveCache();
  }

  search(query) {
    const q = query.toLowerCase();
    const results = new Set();

    for (const [sessionId, entry] of Object.entries(this.cache)) {
      for (const r of entry.resources) {
        if (r.id && r.id.includes(q)) { results.add(sessionId); break; }
        if (r.url && r.url.toLowerCase().includes(q)) { results.add(sessionId); break; }
        if (r.name && r.name.toLowerCase().includes(q)) { results.add(sessionId); break; }
        if (r.repo && r.repo.toLowerCase().includes(q)) { results.add(sessionId); break; }
      }
    }

    return results;
  }
}

module.exports = ResourceIndexer;
module.exports.resourceKey = resourceKey;

module.exports.parseUrlToResource = function parseUrlToResource(url) {
  url = url.trim();
  let m;
  if ((m = url.match(/\/pullrequest\/(\d+)/))) {
    const repoMatch = url.match(/_git\/([^/]+)\/pullrequest/);
    return { type: 'pr', id: m[1], url, repo: repoMatch ? repoMatch[1] : null, state: null };
  }
  if ((m = url.match(/\/_workitems\/edit\/(\d+)/))) {
    return { type: 'workitem', id: m[1], url };
  }
  if ((m = url.match(/\/_build\/results\?buildId=(\d+)/))) {
    return { type: 'pipeline', id: m[1], url };
  }
  if ((m = url.match(/\/_build\?definitionId=(\d+)/))) {
    return { type: 'pipeline', id: `def-${m[1]}`, url };
  }
  if ((m = url.match(/releaseId=(\d+)/))) {
    return { type: 'release', id: m[1], url };
  }
  if (url.match(/_wiki\//)) {
    return { type: 'wiki', url };
  }
  if (url.match(/_git\//)) {
    url = normalizeRepoUrl(url);
    const repoName = url.match(/_git\/(.+)/)?.[1] || url;
    return { type: 'repo', name: repoName, url };
  }
  // Generic link
  return { type: 'link', url, name: url.replace(/^https?:\/\//, '').split('/').slice(0, 3).join('/') };
};
